merge with my branch trackerPatRec

gianipez opened the pull request:
- small bug fixed in the RobustHelixFit algorithm
- switched double -> float in the CalHelixFinder modules/classes
- added new class to evaluate the median from a given set of elements (used only in the RobustHelixFit class)
- minor bug fixes in a few prolog_trigger files
- added CRVFilters directory for filter modules that use the CRV info

brownd1978 commented on TrkReco/src/RobustHelixFit.cc:
```diff
@@ -325,7 +325,7 @@ namespace mu2e
 	if ( facezF1 == facezF2 )  continue;
 
 	float dz = hitP2->pos().z() - hitP1->pos().z();//facezF2->z - facezF1->z;
-	if (dz < _minzsep || dz > _maxzsep)          continue;
```
There is also a statement dz < _minzsep on line 766 of this file, shouldn't that get fixed as well?

gianipez replied:
I guess you are right! I didn't notice it because my students debugged only the initFZ function.

gianipez replied:
Hi Dave,

here you can find a few plots I just made with the CE+mix sample. The blue line is my repository before I made the change you suggested, while the dots represent the results after I made the change.
https://drive.google.com/drive/folders/1vw7Hg76buFwTz3EKAmjS3U3hbUAKEQbe?usp=sharing

Let me know if you want additional plots.

Cheers,

Giani

brownd1978 commented on CaloFilters/src/CaloClusterCounterFilter_module.cc:
```diff
@@ -0,0 +1,155 @@
+  };
+
+
+  CaloClusterCounter::CaloClusterCounter(const fhicl::ParameterSet & pset) :
```
We are requiring validated fcl parameters for all new modules, please update these parameters.

gianipez replied:
can you please point me to an example?
Cheers,

Giani

brownd1978 replied:
MergeHelices_module.cc, SelectRecoMC_module.cc, etc.  Just grep for "struct Config"

gaponenko commented:
Hi Giani,

> can you please point me to an example?

Here are some pointers to get started with configuration validation.

A short Mu2e analyzer module that can be used as an example:
Analyses/src/SimParticleTimeMapAnalyzer_module.cc

Art wiki - sufficient for most simple uses:
https://cdcvs.fnal.gov/redmine/projects/art/wiki/Configuration_validation_and_description

Note the non-obvious typedef requirement in the "Enabling description
printouts" section.  Without it the validation works, but the
documentation does not show up in "mu2e --print-description".

There are other wikis with more details:

FHICl wiki:
https://cdcvs.fnal.gov/redmine/projects/fhicl-cpp/wiki/Configuration_validation_and_fhiclcpp_types

And a page with information that was hard to find, about converting
FHICL into "nearly arbitrary C++ type":
https://cdcvs.fnal.gov/redmine/projects/fhicl-cpp/wiki/Auxiliary_classes#OptionalTupleAs

kutschke commented:
I have a lot of requests for small changes on Mu2eUtilities/inc/MedianCalculator.hh.  After lunch and the Ops Planning meeting I will figure out how to request changes and to use the commenting on the code features.

gianipez commented:
Hi Rob,

my answers follow. 

> I will start with a few questions on how this code will be used and then move on to more technical questions
> 
> 1. I presume that the function argument of type std::vector is accumulated in the calling code.  Is it used in the calling code for other purposes or just as input to the call to MedianCalculator?

The vector(s) are used only to evaluate the median. In the code (RobustHelixFit.cc) you can see that we use a std::vector for each of the quantity we reconstruct. 

> 2. In the way that this class is used, can there be two subsequent calls to MedianCalculator with the content of the vector unchanged? If so, you are needlessly resorting.

I am not sure I got the point of this question... in the current structure of the code, every time we use the medianCalculator the input is different, so the sorting is done once per vector.

> 3. The function MedianCalculator has the side effect that it modifies it's input argument.  Is the information that the vector is now sorted used by the calling code?

No, these vectors are used only for the median calculation.

> 4. Remove the lines
>    // $Id: $
>    // $Author: $
>    // $Date: $
>    These lines are present in some old files that were present when we used cvs, not git, as our source code management system. The information after the : and before the trailing $ was filled in by cvs.  They have no value with git and we plan to remove all of them.

OK

> 5. The implementation of  extractMedian should be in a .cc file, not in a .hh file.  It just wastes compile time and memory.  In general the only functions that should be considered for inlining are things like accessors and comparison operators that will complile to code that is:
>    a) smaller than the corresponding function call
>    b) faster at run time than the corresponding function call
>    Of course template implementations need to be in a .hh file.

OK, I will create the corresponding cc file.

> 6. Suppose that typical values are large numbers, or negative numbers; in those cases your code will silently return a grossly incorrect answer in the case of zero entries.  Is there a good reason not to throw an exception in the case of zero entries?

I can certainlt add it. Thanks for the suggestion.

> 7. The class MedianCalculator has no state and only one function that could have been written as a free function.  Why make it a class and not a free function?

I was wondering about possible improvements in the future with new functions (like, a plain median caclulator), or a possible need functions for debugging. I would be also comfortable in making it a free function if the reviewers believe is a better design choice.

> 8. I think you can choose a better name than MedianCalculator
> 
> * it's a weighted mean, not just a plain mean
> * it hides the important side effect that it modifies its argument
>   My recommended name depends on the answers to 1,2,3 and 9.
> 
> 1. One of my pet peeves is names line MedianDatacomp.  It's never really exposed to the end user and the scope is small - this is not too important. I would have called the struct LessByValue, with the argument types distinguishing what I am comparing.
> 2. You should consider putting MedianData and MedianDatacomp into an anonymous namespace.  They really have no use outside of this narrow scope; since they are very generic names you should hide them from other namespaces.

OK


I will wait suggestions for a new name before commiting the new version.

Thanks for the useful comments,

Giani

kutschke commented:
Hi Giani,

  I am replying by email rather than logging in to github - let's see how this works.

  Your code is a great but minimalist solution to the immediate needs of your algorithm.  But I think it's too minimalist to promote it as a general purpose tool.  I can provide details if you want.

  The big decision that I see is that we do one of 2 things:

1) We leave this more or less as it is with the small changes I discussed earlier and we should
     ask if it still belongs in Mu2eUtilities - what do others think?

2) Replace what you have with a general purpose  class that does both weighted mean and median.
    We an add higher order moments if needed.  It's not clear if this would be significantly faster or slower -
    there is the cost of accumulating sums for the mean but you can remove a loop from the implementation.

If you are interested in 2) I can give you details - it's only a minor amount of extra work.

Two other things to fix:

1) A trivial test program like:
#include "Mu2eUtilities/inc/MedianCalculator.hh"
void foo (){}
does not compile.  Your code relies on something that is getting included via something
#included in RobustHelixFit.cc   Please fix this.

2)
I bet that most of the #includes in MedianCalculator.hh are not needed.
Please remove the ones that are not needed.

A few answers to your questions below.


  Rob


> On Aug 30, 2019, at 6:25 PM, Gianantonio Pezzullo <notifications@github.com> wrote:
> 
> Hi Rob,
> 
> my answers follow.
> 
> I will start with a few questions on how this code will be used and then move on to more technical questions
> 
> 	• I presume that the function argument of type std::vector is accumulated in the calling code. Is it used in the calling code for other purposes or just as input to the call to MedianCalculator?
> The vector(s) are used only to evaluate the median. In the code (RobustHelixFit.cc) you can see that we use a std::vector for each of the quantity we reconstruct.

In that case it would be reasonable for the main class to hold the std::vector<MedianData> as member data.
To my mind this choice simplifies the code in the calling function - the calling code has a single object rather than multiple cooperating objects.  

By the way, I really like the fact that you defined the struct MedianData and did not use std::pair<float,float>.

> 
> 	• In the way that this class is used, can there be two subsequent calls to MedianCalculator with the content of the vector unchanged? If so, you are needlessly resorting.
> I am not sure I got the point of this question... in the current structure of the code, every time we use the medianCalculator the input is different, so the sorting is done once per vector.

I was thinking ahead the general purpose solution in which the main class holds the std::vector.  In a general use case (not your use case), it would be important not to redo the sort and search if it were unnecessary.

> 	• The function MedianCalculator has the side effect that it modifies it's input argument. Is the information that the vector is now sorted used by the calling code?
> No, these vectors are used only for the median calculation.

Ok - another point in favor of putting the vector inside the main class.  A general user who wants to use your code might be very surprised that computing the median invalidates all iterators into their std::vector.

> 	• The class MedianCalculator has no state and only one function that could have been written as a free function. Why make it a class and not a free function?
> I was wondering about possible improvements in the future with new functions (like, a plain median caclulator), or a possible need functions for debugging. I would be also comfortable in making it a free function if the reviewers believe is a better design choice.

If we want to keep this minimalist, just make it a free function.  If we want more features we should plan them now.

> 	• You should consider putting MedianData and MedianDatacomp into an anonymous namespace. They really have no use outside of this narrow scope; since they are very generic names you should hide them from other namespaces

I take this one back.  The anonymous namespace won't work for MedianData, which needs to stay in the .hh file.  And MedianDatacomp should be defined in the same scope as MedianData.

> 
> I will wait suggestions for a new name before commiting the new version.

Let's make the big decision first and the rest will be fast.





> 
> Thanks for the useful comments,
> 
> Giani
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

gianipez commented:
Hi Rob et al,

I agree that adding the vector to the "**MedianCalculator**" class makes things much cleaner. I implemented it and I added a function that calculates the unweighted median. I organized the code in a way that sorting and median calculations are made only if the vector is increased. Please, have a look. I also started cleaning up the **SConscript** in **CRVFilters/src**, but I am having some difficulties in using the option -Wl,--as-needed (I sent you a private email). 
I also re-aligned my code with master yesterday night.

Please, let me know how do you want to proceed.

Best,

Giani

kutschke commented on CRVFilters/src/CrvCoincidenceClusterFilter_module.cc:
```diff
@@ -0,0 +1,152 @@
+	Comment("label of the given trigger-path")};
+    };
+
+    enum {
```
I am trying this comment again, this time properly tied to the line number.   The enum defined on line 74 of CRVFilters/src/CrvCoincidenceClusterFilter_module.cc  is not used elsewhere in the code.  Please delete it.

gianipez commented:
OK, I will do it ASAP. I had hard time tracing back the lines of that specific comment.

Sorry about that,


Gianantonio Pezzullo, Ph.D.
giani.pezzu@gmail.com


**********************************************************************************************************
Address: Department of Physics, Yale University, 56 Hillhouse, New Haven, CT-06511, USA
						     		  
**********************************************************************************************************

> On Sep 3, 2019, at 12:16 PM, Rob Kutschke <notifications@github.com> wrote:
> 
> @kutschke commented on this pull request.
> 
> In CRVFilters/src/CrvCoincidenceClusterFilter_module.cc <https://github.com/Mu2e/Offline/pull/4#discussion_r320360060>:
> 
> > +     
> +  public:
> +    using Name=fhicl::Name;
> +    using Comment=fhicl::Comment;
> +    struct Config {
> +      fhicl::Atom<int> diag{ Name("ddiagLevel"),
> +	Comment("Diagnostic Level"), 0};
> +      fhicl::Atom<art::InputTag> CRVCCF{ Name("CrvCoincidenceClusterFinder"),
> +	Comment("CrvCoincidenceClusterFinder producer")};
> +      fhicl::Atom<int> minNCC{ Name("MinNCC"),
> +	Comment("Minimum number of cluster of coincidences")};
> +      fhicl::Atom<std::string> trgPath{ Name("triggerPath"),
> +	Comment("label of the given trigger-path")};
> +    };
> +
> +    enum {
> I am trying this comment again, this time properly tied to the line number. The enum defined on line 74 of CRVFilters/src/CrvCoincidenceClusterFilter_module.cc is not used elsewhere in the code. Please delete it.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Mu2e/Offline/pull/4?email_source=notifications&email_token=AGCD3EC3BLMKAO42MPXZYJ3QH2EWDA5CNFSM4ISOHKV2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDP6S6Q#pullrequestreview-283109754>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AGCD3EHRGKLVXO5NIWV2DCTQH2EWDANCNFSM4ISOHKVQ>.
>

brownd1978 commented:
Giani, I've gotten lost on this thread.  Can you please summarize the status of:
- abs(dz) in RobustHelixFinder
- Offset in Lambda in CalHelixFinder
- validated fhicl in CaloClusterCounter
thanks, Dave

gianipez commented:
Hi Dave,

my comments below:

> On Sep 3, 2019, at 2:54 PM, David Brown <notifications@github.com> wrote:
> 
> Giani, I've gotten lost on this thread. Can you please summarize the status of:
> 
> abs(dz) in RobustHelixFinder
Done and tested (I put the plots in the google drive)

> Offset in Lambda in CalHelixFinder
working on it. Do you want me to fix it before this pull or can it be a separate one?

> validated fhicl in CaloClusterCounter

doing it right now (I don’t know why I implemented the changes only in the CRV filter), sorry about that.
> thanks, Dave
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Mu2e/Offline/pull/4?email_source=notifications&email_token=AGCD3EHVTGEJLZJB5FN326TQH2XE5A5CNFSM4ISOHKV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5ZGNKA#issuecomment-527591080>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AGCD3EEMK5YMSTUQAK56KRLQH2XE5ANCNFSM4ISOHKVQ>.
>

brownd1978 commented:
Hi Giani,

On Tue, Sep 3, 2019 at 12:18 PM Gianantonio Pezzullo <
notifications@github.com> wrote:

> Hi Dave,
>
> my comments below:
>
> > On Sep 3, 2019, at 2:54 PM, David Brown <notifications@github.com>
> wrote:
> >
> > Giani, I've gotten lost on this thread. Can you please summarize the
> status of:
> >
> > abs(dz) in RobustHelixFinder
> Done and tested (I put the plots in the google drive)
>

The plots all look good: I signoff on this part of your commit.


>
> > Offset in Lambda in CalHelixFinder
> working on it. Do you want me to fix it before this pull or can it be a
> separate one?
>

This can be separate, I just wanted to know your plans.


>
> > validated fhicl in CaloClusterCounter
>
> doing it right now (I don’t know why I implemented the changes only in the
> CRV filter), sorry about that.
>

OK thanks.  I'll look for your commit and will approve once I see it.

cheers, Dave


> > thanks, Dave
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub <
> https://github.com/Mu2e/Offline/pull/4?email_source=notifications&email_token=AGCD3EHVTGEJLZJB5FN326TQH2XE5A5CNFSM4ISOHKV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5ZGNKA#issuecomment-527591080>,
> or mute the thread <
> https://github.com/notifications/unsubscribe-auth/AGCD3EEMK5YMSTUQAK56KRLQH2XE5ANCNFSM4ISOHKVQ
> >.
> >
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/4?email_source=notifications&email_token=ABAH572CH7P2UQEC2YPNHH3QH22B5A5CNFSM4ISOHKV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5ZIZUA#issuecomment-527600848>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABAH576ZGLKX2DKS2VFKAE3QH22B5ANCNFSM4ISOHKVQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gianipez commented:
Dear reviewers,

I believe I implemented all the changes you aked me. Please, let me know if you want additional changes.

Cheers,

Giani

gianipez commented:
Hi Rob,

I think I completed all the changes you requested. 

Thanks for the comments,

Giani
